File: src/morie/fn/cntgc.py
# morie.fn -- function file (rootcoder007/morie)
"""Pearson contingency coefficient C (Gibbons Ch 14.2.1).

C = sqrt(chi2 / (chi2 + n))

Also reports Cramer's V = sqrt(chi2 / (n * (min(r, c) - 1))) and
the maximum attainable C = sqrt((min(r,c)-1)/min(r,c)) so users
can normalise.
"""
from __future__ import annotations

import numpy as np
from scipy import stats

from ._richresult import RichResult

__all__ = ["contingency_coefficient"]


def contingency_coefficient(x):
    """Pearson contingency coefficient and friends.

    Parameters
    ----------
    x : 2-D array-like
        Two-way contingency table of counts (rows x cols).

    Returns
    -------
    RichResult with payload:
        statistic   : Pearson's C
        cramers_v   : Cramer's V (chi-square based)
        chi2        : Pearson chi-square statistic
        p_value     : chi-square p-value (df = (r-1)(c-1))
        df          : degrees of freedom
        max_C       : theoretical upper bound for C given table size
        n           : grand total
    """
    X = np.asarray(x, dtype=float)
    if X.ndim != 2 or X.size == 0:
        return RichResult(payload={
            "statistic": np.nan, "cramers_v": np.nan,
            "chi2": np.nan, "p_value": np.nan, "df": np.nan,
            "max_C": np.nan, "n": 0,
            "method": "Pearson contingency coefficient",
        })
    chi2, p, df, _ = stats.chi2_contingency(X, correction=False)
    n_total = float(X.sum())
    C = np.sqrt(chi2 / (chi2 + n_total)) if n_total > 0 else np.nan
    r, c = X.shape
    mn = min(r, c)
    V = np.sqrt(chi2 / (n_total * (mn - 1))) if (n_total > 0 and mn > 1) else np.nan
    max_C = np.sqrt((mn - 1) / mn) if mn > 1 else np.nan
    return RichResult(payload={
        "statistic": float(C),
        "cramers_v": float(V),
        "chi2": float(chi2),
        "p_value": float(p),
        "df": int(df),
        "max_C": float(max_C),
        "n": int(n_total),
        "method": "Pearson contingency coefficient",
    })


def cheatsheet():
    return "cntgc: Pearson contingency coefficient C"


# CANONICAL TEST
# >>> contingency_coefficient([[10, 0], [0, 10]])
# Perfect 2x2 association: chi2 = 20, C = sqrt(20/30) = 0.8165

File: r-package/morie/R/cntgc.R
# SPDX-License-Identifier: MIT OR Apache-2.0

#' Pearson contingency coefficient C (Gibbons Ch 14.2.1)
#'
#' C = sqrt(chi^2 / (chi^2 + n)).  Also reports Cramer's V and the
#' maximum attainable C = sqrt((min(r,c)-1)/min(r,c)).
#'
#' @param x A 2-D contingency table of counts.
#' @return Named list: statistic (C), cramers_v, chi2, p_value, df,
#'   max_C, n.
#' @importFrom stats chisq.test
#' @examples
#' \dontrun{
#'   # See the package vignettes for usage examples:
#'   #   vignette(package = "morie")
#' }
#' @export
contingency_coefficient <- function(x) {
  X <- as.matrix(x)
  if (length(dim(X)) != 2L || length(X) == 0L) {
    return(list(statistic = NA_real_, cramers_v = NA_real_,
                chi2 = NA_real_, p_value = NA_real_, df = NA_integer_,
                max_C = NA_real_, n = 0L,
                method = "Pearson contingency coefficient"))
  }
  ct <- suppressWarnings(stats::chisq.test(X, correct = FALSE))
  n_total <- sum(X)
  chi2 <- as.numeric(ct$statistic)
  C <- sqrt(chi2 / (chi2 + n_total))
  r <- nrow(X); c <- ncol(X); mn <- min(r, c)
  V <- if (mn > 1) sqrt(chi2 / (n_total * (mn - 1))) else NA_real_
  max_C <- if (mn > 1) sqrt((mn - 1) / mn) else NA_real_
  list(
    statistic = C,
    cramers_v = V,
    chi2 = chi2,
    p_value = as.numeric(ct$p.value),
    df = as.integer(ct$parameter),
    max_C = max_C,
    n = as.integer(n_total),
    method = "Pearson contingency coefficient"
  )
}
